webdataset prototype - LoadFilesFromDiskIterableDataset

glaringlee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49353 webdataset prototype - GroupByFilenameIterableDataset
* #49140 webdataset prototype - RoutedDecoderIterableDataset
* #49026 webdataset prototype - ReadFilesFromZipIterableDataset
* #49000 webdataset prototype - ReadFilesFromTarIterableDataset
* **#48955 webdataset prototype - LoadFilesFromDiskIterableDataset**
* #48944 webdataset prototype - ListDirFilesIterableDataset

Differential Revision: [D25541393](https://our.internmc.facebook.com/intern/diff/D25541393)

Commit:
webdataset prototype - LoadFilesFromDiskIterableDataset

[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - LoadFilesFromDiskIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - LoadFilesFromDiskIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - LoadFilesFromDiskIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - LoadFilesFromDiskIterableDataset"




[ghstack-poisoned]

VitalyFedyunin commented on test/test_dataset.py:
```diff
@@ -3,14 +3,15 @@
-    temp_file3 = tempfile.NamedTemporaryFile(dir=temp_dir_path, delete=False)
+    temp_file1 = tempfile.NamedTemporaryFile(dir=temp_dir_path, delete=False)  # noqa: P201
+    temp_file2 = tempfile.NamedTemporaryFile(dir=temp_dir_path, delete=False)  # noqa: P201
+    temp_file3 = tempfile.NamedTemporaryFile(dir=temp_dir_path, delete=False)  # noqa: P201
```
Can you please commend this noqa

glaringlee replied:
These #noqa: P201 is removed in the later stack PRs, i changed another way to create these files.

VitalyFedyunin commented on test/test_dataset.py:
```diff
@@ -34,5 +35,14 @@ def test_listdirfiles_iterable_dataset(self):
+        dataset1 = ListDirFilesIterableDataset(temp_dir, '')
+        dataset2 = LoadFilesFromDiskIterableDataset(dataset1)
+
+        for rec in dataset2:
```
Are you sure that traverse order is guaranteed?

glaringlee replied:
Traverse order is guaranteed in LoadFilesFromDiskIterableDataset.
I think the question is whether the traverse order is guaranteed in ListDirFilesIterableDataset.
ListDirFilesIterableDataset use os.walk, currently no enforced order, should we enforce some kind of order? sort by file path+name?

VitalyFedyunin replied:
No need to order it.

VitalyFedyunin commented on torch/utils/data/datasets/common.py:
```diff
@@ -35,3 +35,17 @@ def onerror(err : OSError):
+            warnings.warn("file pathname must be string type, but got {}".format(type(pathname)))
+            raise TypeError
+
+        yield (pathname, open(pathname, 'rb'))
```
Can you please verify that we are releasing file handler with created stream objects gets garbage collected

glaringlee replied:
Ah, good catch. I don't think all the platform will do this releasing automatically. I will update the consumer dataset (tardataset, decoder dataset, etc) to try explicitly release it if possible.

VitalyFedyunin replied:
Alternatively wrap (or find wrapper) which will take care of closing stream on object __del__

glaringlee replied:
Double checked the impl of BufferedIOBased class (which is used as streaming handle obj), it inherited from IOBased which has __del__ implemented calling close() already, we don't need to take care of this manually.

Commit:
Update on "[WIP] webdataset prototype - LoadFilesFromDiskIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - LoadFilesFromDiskIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - LoadFilesFromDiskIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - LoadFilesFromDiskIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - LoadFilesFromDiskIterableDataset"




[ghstack-poisoned]

VitalyFedyunin commented on torch/utils/data/datasets/common.py:
```diff
@@ -2,6 +2,41 @@
 from typing import List, Union, Iterable
+from io import BufferedIOBase
+
+class StreamWrapper:
```
Need to be Child of Stream class or similar, to make sure that isinstance work correctly.

glaringlee replied:
Double checked the BufferedIOBase which is used as streaming handle. It inherited from IOBase which has __del__ implemented calling close() already. I'll remove this StreamWrapper, no longer needed.

Commit:
Update on "[WIP] webdataset prototype - LoadFilesFromDiskIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - LoadFilesFromDiskIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - LoadFilesFromDiskIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - LoadFilesFromDiskIterableDataset"




[ghstack-poisoned]

Commit:
Update on "webdataset prototype - LoadFilesFromDiskIterableDataset"


Differential Revision: [D25541393](https://our.internmc.facebook.com/intern/diff/D25541393)

[ghstack-poisoned]

Commit:
Update on "webdataset prototype - LoadFilesFromDiskIterableDataset"


Differential Revision: [D25541393](https://our.internmc.facebook.com/intern/diff/D25541393)

[ghstack-poisoned]

Commit:
Update on "webdataset prototype - LoadFilesFromDiskIterableDataset"


Differential Revision: [D25541393](https://our.internmc.facebook.com/intern/diff/D25541393)

[ghstack-poisoned]

Commit:
Update on "webdataset prototype - LoadFilesFromDiskIterableDataset"


Differential Revision: [D25541393](https://our.internmc.facebook.com/intern/diff/D25541393)

[ghstack-poisoned]

Commit:
Update on "webdataset prototype - LoadFilesFromDiskIterableDataset"


Differential Revision: [D25541393](https://our.internmc.facebook.com/intern/diff/D25541393)

[ghstack-poisoned]

Commit:
Update on "webdataset prototype - LoadFilesFromDiskIterableDataset"


Differential Revision: [D25541393](https://our.internmc.facebook.com/intern/diff/D25541393)

[ghstack-poisoned]
